backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/versions/CassandraSchemaVersionDAO.java

```diff
@@ -51,27 +55,28 @@ public CassandraSchemaVersionDAO(Session session) {
 
     private PreparedStatement prepareReadVersionStatement(Session session) {
         return session.prepare(
-            QueryBuilder.select(VALUE)
-                .from(TABLE_NAME)
-                .where(QueryBuilder.eq(KEY, KEY_FOR_VERSION)));
+            select(VALUE)
+                .from(TABLE_NAME));
     }
 
     private PreparedStatement prepareWriteVersionStatement(Session session) {
         return session.prepare(
-            QueryBuilder.insertInto(CassandraSchemaVersionTable.TABLE_NAME).value(KEY, KEY_FOR_VERSION)
+            insertInto(CassandraSchemaVersionTable.TABLE_NAME)
+                .value(KEY, bindMarker(KEY))
                 .value(VALUE, bindMarker(VALUE)));
     }
 
     public CompletableFuture<Optional<Integer>> getCurrentSchemaVersion() {
-        return cassandraAsyncExecutor.executeSingleRow(readVersionStatement.bind())
-            .thenApply(rowOptional ->
-                rowOptional
-                    .map(row -> row.getInt(VALUE)));
+        return cassandraAsyncExecutor.execute(readVersionStatement.bind())
+            .thenApply(resultSet -> CassandraUtils.convertToStream(resultSet)
+                .map(row -> row.getInt(VALUE))
+                .reduce(Math::max));
```

Comment:
Downgrade is not supported.
It's a new migration where v1 schema = v3 schema (v1 -> v2 -> v3).